### PR TITLE
Telegram Integration

### PR DESCRIPTION
**Pull request by abhaymathur21**

### CODE REVIEW

File: jb-lib/lib/channel_handler/__init__.py [hunk 1]
```diff
@@ -0,0 +1,2 @@
+from lib.channel_handler.channel_handler import ChannelHandler, ChannelData, User
+from lib.channel_handler.pinnacle_whatsapp_handler import PinnacleWhatsappHandler
```
Comment (shreypandey) [inline jb-lib/lib/channel_handler/__init__.py:2]: Import telegram Handler here

Comment (shreypandey) [reply jb-lib/lib/channel_handler/__init__.py:2]: Convert lib.channel_handler to . convention

File: jb-lib/lib/channel_handler/pinnacle_whatsapp_handler.py [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+    ListMessage,
+    ButtonMessage,
+    DialogMessage,
+    DialogOption,
+    FormReplyMessage
+)
+from lib.models import JBChannel, JBUser, JBForm
+from lib.utils import EncryptionHandler
+
+azure_creds = {
```
Comment (shreypandey) [inline jb-lib/lib/channel_handler/pinnacle_whatsapp_handler.py:34]: Use lib.file_storage instead.

Refer src/extensions.py in language

File: jb-lib/lib/channel_handler/telegram_handler.py [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+    ListMessage,
+    ButtonMessage,
+    DialogMessage,
+    DialogOption,
+    FormReplyMessage
+)
+from lib.models import JBChannel, JBUser, JBForm
+from lib.utils import EncryptionHandler
+
+azure_creds = {
```
Comment (shreypandey) [inline jb-lib/lib/channel_handler/telegram_handler.py:34]: Same as before
Use file_storage

File: jb-lib/lib/channel_handler/telegram_handler.py [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+    "base_path": "input/",
+}
+storage = AzureStorageSync(**azure_creds)
+
+
+class TelegramHandler(RestChannelHandler):
+
+    @classmethod
+    def is_valid_data(cls, data: Dict) -> bool:
+        return "message" in data or "edited_message" in data
```
Comment (shreypandey) [inline jb-lib/lib/channel_handler/telegram_handler.py:47]: Is there any better way to identify if the message is from telegram?

File: jb-lib/lib/channel_handler/telegram_handler.py [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+        decrypted_key = EncryptionHandler.decrypt_text(encryption_key)
+        headers = {
+            "Content-type": "application/json",
+            "Authorization": f"Bearer {decrypted_key}",
+        }
+        return headers
+
+    @classmethod
+    def send_message(cls, channel: JBChannel, user: JBUser, message: Message):
+        url = f'https://api.telegram.org/bot{channel.app_id}/sendMessage'
```
Comment (shreypandey) [inline jb-lib/lib/channel_handler/telegram_handler.py:354]: Fetch this URL from JBChannel

File: jb-lib/lib/channel_handler/__init__.py (hunk 1 above)
Comment (abhaymathur21) [reply jb-lib/lib/channel_handler/__init__.py]: done

File: jb-lib/lib/channel_handler/pinnacle_whatsapp_handler.py (hunk 2 above)
Comment (abhaymathur21) [reply jb-lib/lib/channel_handler/pinnacle_whatsapp_handler.py]: done

File: jb-lib/lib/channel_handler/telegram_handler.py (hunk 3 above)
Comment (abhaymathur21) [reply jb-lib/lib/channel_handler/telegram_handler.py]: done

File: jb-lib/lib/channel_handler/telegram_handler.py (hunk 5 above)
Comment (abhaymathur21) [reply jb-lib/lib/channel_handler/telegram_handler.py]: even in the whatsapp handler?

File: jb-lib/lib/channel_handler/telegram_handler.py (hunk 4 above)
Comment (abhaymathur21) [reply jb-lib/lib/channel_handler/telegram_handler.py:41]: fixed

File: jb-lib/lib/channel_handler/telegram_handler.py (hunk 5 above)
Comment (shreypandey) [reply jb-lib/lib/channel_handler/telegram_handler.py]: @abhaymathur21 Yes

File: jb-lib/lib/channel_handler/channel_handler.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        pass
+
+    @classmethod
+    @abstractmethod
+    def get_channel_name(cls) -> str:
+        pass
+
+    @classmethod
+    @abstractmethod
+    def to_message(cls, turn_id: str, channel: JBChannel, bot_input: BotInput) -> Message:
```
Comment (sameersegal) [inline jb-lib/lib/channel_handler/channel_handler.py:35]: Why turn_id?

File: jb-lib/lib/channel_handler/channel_handler.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    user_phone: Optional[str] = None
+
+
+class ChannelData(BaseModel):
+    bot_identifier: str
+    user: User
+    message_data: Dict
+
+
+class ChannelHandler(ABC):
```
Comment (sameersegal) [inline jb-lib/lib/channel_handler/channel_handler.py:21]: Write docstrings for each of these functions so that we are clear what their scope is

**Review by sameersegal — CHANGES_REQUESTED** [review]
Describe your changes in brief text (and code wherever relevant) to explain the overall change.

File: api/tests/test_activate_bot.py [hunk 8]
```diff
@@ -2,7 +2,19 @@
 import pytest
 from app.jb_schema import JBBotActivate, JBBotChannels
 from lib.models import JBBot, JBChannel
-from app.handlers.v1.bot_handlers import handle_activate_bot
+
+mock_sync_storage_instance = MagicMock()
```
Comment (sameersegal) [inline api/tests/test_activate_bot.py:6]: These should not be there

File: api/tests/test_callback.py [hunk 9]
```diff
@@ -6,7 +6,19 @@
     ChannelIntent,
 )
 
-from app.handlers.v1 import handle_callback
+mock_sync_storage_instance = MagicMock()
```
Comment (sameersegal) [inline api/tests/test_callback.py:9]: same here

**Comment by KaranrajM** [comment]
Hi @shreypandey if storage was to be used in api service, the below code can be taken for reference

from lib.file_storage import StorageHandler
storage = StorageHandler.get_async_instance()
await storage.write_file(
    file_path=file.filename,
    file_content=file.file_content,
    mime_type=file.content_type,
)
file_path = await self.storage._download_file_to_temp_storage(
    file.filename
)

Then the list of file_path are sent to indexer for reading and indexing them.

File: jb-lib/lib/channel_handler/__init__.py [hunk 10]
```diff
@@ -0,0 +1,9 @@
+from typing import Dict
+from .channel_handler import ChannelHandler, ChannelData, User
+from .pinnacle_whatsapp_handler import PinnacleWhatsappHandler
+from .telegram_handler import TelegramHandler
+
+channel_map: Dict[str, type[ChannelHandler]] = {
```
[the remark this answers is not in the document]
Comment (shreypandey) [reply jb-lib/lib/channel_handler/__init__.py:6]: We have the following requirements:
1. Get the list of all available channels
2. Iterate over all channels to find the relevant channel for a given payload

File: docker-compose.yml [hunk 11]
```diff
@@ -16,6 +16,11 @@ services:
       - KAFKA_FLOW_TOPIC=${KAFKA_FLOW_TOPIC}
       - ENCRYPTION_KEY=${ENCRYPTION_KEY}
       - WA_API_HOST=${WA_API_HOST}
+      - STORAGE_TYPE=${STORAGE_TYPE}
```
[the remark this answers is not in the document]
Comment (shreypandey) [reply docker-compose.yml]: This is fixed

**Comment by shreypandey** [comment]
@sameersegal I have refactored the channel code for incoming message as follows:
```python
async def process_incoming_messages(turn_id: str, bot_input: RestBotInput):
    """Process incoming messages"""
    channel = channel_map[bot_input.channel_name]
    jb_channel = await get_channel_by_turn_id(turn_id)

    message_type = channel.get_message_type(bot_input)
    logger.info("Message type: %s", message_type)

    if message_type == MessageType.TEXT:
        text_message = channel.to_text_message(bot_input)
        message = Message(
            message_type=MessageType.TEXT,
            text=text_message,
        )
    elif message_type == MessageType.AUDIO:
        encoded_audio_content = channel.get_audio(jb_channel, bot_input)
        audio_content = base64.b64decode(encoded_audio_content)
        audio_file_name = f"{turn_id}.ogg"
        storage = StorageHandler.get_async_instance()
        await storage.write_file(audio_file_name, audio_content, "audio/ogg")
        storage_url = await storage.public_url(audio_file_name)
        audio_message = AudioMessage(
            media_url=storage_url,
        )
        message = Message(
            message_type=MessageType.AUDIO,
            audio=audio_message,
        )
    elif message_type == MessageType.INTERACTIVE_REPLY:
        interactive_reply_message = channel.to_interactive_reply_message(bot_input)
        message = Message(
            message_type=MessageType.INTERACTIVE_REPLY,
            interactive_reply=interactive_reply_message,
        )
    elif message_type == MessageType.FORM_REPLY:
        form_reply_message = channel.to_form_reply_message(bot_input)
        message = Message(
            message_type=MessageType.FORM_REPLY,
            form_reply=form_reply_message,
        )
    elif message_type == MessageType.DIALOG:
        dialog_message = channel.to_dialog_message(bot_input)
        message = Message(
            message_type=MessageType.DIALOG,
            dialog=dialog_message,
        )
    else:
        return NotImplemented
```


Please let me know if this works.

File: api/tests/test_activate_bot.py (hunk 8 above)
Comment (shreypandey) [reply api/tests/test_activate_bot.py]: This is fixed

File: api/tests/test_callback.py (hunk 9 above)
Comment (shreypandey) [reply api/tests/test_callback.py]: This is fixed

File: jb-lib/lib/channel_handler/channel_handler.py (hunk 6 above)
Comment (shreypandey) [reply jb-lib/lib/channel_handler/channel_handler.py]: This is fixed

File: jb-lib/lib/channel_handler/channel_handler.py (hunk 7 above)
Comment (shreypandey) [reply jb-lib/lib/channel_handler/channel_handler.py:33]: Added

**Review by sameersegal — APPROVED** [review]

**Pull request merged**